[WIP] Add functionality to support AdaMix

rishabbala opened the pull request:
Should close issue #504. Work in Progress

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_719). All of your documentation changes will be reflected on that endpoint.

rishabbala commented:
@younesbelkada I think the PR is tested and working. All tests pass except a test in  `test_gpu_examples` for lora (I haven't changed anything in the lora files, so I'm not sure if this test fail case is from my end. Let me know if this has to be verified). I also tried the sequence to sequence modelling similar to the examples provided [here](https://github.com/huggingface/peft/blob/main/examples/conditional_generation/peft_ia3_seq2seq.ipynb), and get 93% accuracy. One downside is that for larger models the accuracy is much lower (~75%), but this is expected and discussed in their paper (larger model + lesser data works poorly). Overall I think the current code works, but might not be in the best state to be merged. Let me know what changes are required, and I'll also fix the docs for AdaMix

BenjaminBossan commented:
Thanks @rishabbala for the PR. Could you please run `make style` on your code and push the changes? Then the quality check will pass and the tests will run.

rishabbala commented:
Hey @BenjaminBossan I ran `make style`, but there is still some failures. Could you let me know what the issue is

BenjaminBossan commented:
That's strange, locally `make style` did not complain?

Anyway, the issue is that `ruff` complains about:

> src/peft/peft_model.py:696:9: E722 Do not use bare `except`
> tests/testing_common.py:498:21: F631 Assert test is a non-empty tuple, which is always `True`
> tests/testing_common.py:638:21: F631 Assert test is a non-empty tuple, which is always `True`

You can check that when you click on `Details` for the failed `tests / check_code_quality (pull_request)`. Maybe your local `ruff`  version is not the most recent? Could you try:

```
python -m pip install -U ruff
make style
```
If that doesn't help, maybe you can try fixing the quoted issues manually.

rishabbala commented:
Thanks that fixed it

BenjaminBossan commented:
@rishabbala Some tests are failing, could you please investigate?

younesbelkada reviewed (commented):
Thanks a lot for your great work @rishabbala - indeed some CI tests are failing, it would be great if you can have a look 🙏 Afterwards we will be able to review the PR and discuss the next steps before merging this ! 
Let us know if you need any help, thanks!

rishabbala commented:
Hi @younesbelkada, the error is because of an in-place modification of the embeddings in the deberta and deberta v2 models (line 795 in deberta `embeddings += position_embeddings`). Locally I was able to fix this by changing this line in both models to `embeddings = embeddings + position_embeddings`. However, the current fix in the peft repository is to skip testing for deberta models for all PEFT methods implemented so far (line 122 tests/test_feature_extraction_models.py - `filter_params_func=skip_deberta_lora_tests`). I think this might not be the best way to address this issue, and would propose modifying the files in huggingface transformers, but if you feel that it would be better to skip these tests, I can do that and push the code again

BenjaminBossan commented:
> but if you feel that it would be better to skip these tests, I can do that and push the code again

Yes, let's skip those tests for now. The issue you mention might be worth bringing up in transformers.

rishabbala commented:
@BenjaminBossan Done, could you run the tests again

BenjaminBossan commented:
@rishabbala Could you please fix the merge conflict, then we can run the CI.

rishabbala commented:
@BenjaminBossan the conflicts are resolved

BenjaminBossan commented on tests/testing_common.py:
```diff
@@ -38,13 +39,22 @@
+        "sharing_down": False,
+        "sharing_up": True,
+        "return_two_views": True,
+    },
```
Can we get a 2nd config with slightly different params, e.g.:

```
    {
        "target_modules": None,
        "adapter_dim": 8,
        "num_expert": 3,
        "sharing_down": True,
        "sharing_up": False,
        "return_two_views": False,
    },
```

That way, we get better test coverage.

BenjaminBossan commented on src/peft/tuners/adamix.py:
```diff
@@ -0,0 +1,416 @@
+            result1 = self.MoA_up(result1)
+            result1 = result1 + x
+
+            if self.training and self.return_two_views:
```
Is it possible that `result2` would just use the exact same expert, leading to 0 divergence? If so, is this consistent with the paper? Maybe it doesn't matter in the grand scheme of things.

rishabbala replied:
Thats possible, and is consistent with the paper.

BenjaminBossan commented on src/peft/tuners/adamix.py:
```diff
@@ -0,0 +1,416 @@
+            expert_idx = torch.randint(low=0, high=self.num_local_experts, size=(1,)).item()  # selected expert
+            expert_output = self.get_expert_by_idx(expert_idx)(x)
+
+        else:
```
According to test coverage, these lines are never tested, i.e. there are no tests for `inference_mode`. Could you please add some?

rishabbala replied:
Changed this to be triggered when we do model.eval()

BenjaminBossan commented on src/peft/tuners/adamix.py:
```diff
@@ -0,0 +1,416 @@
+            )
+
+    @staticmethod
+    def _is_valid_match(key: str, target_key: str):
```
If I'm not mistaken, this method is not used anywhere and can be deleted.

BenjaminBossan commented on src/peft/tuners/adamix.py:
```diff
@@ -0,0 +1,416 @@
+        return target_module_found
+
+    def get_param_dtype(self, target):
+        params = list(target.parameters())
```
Could this be simplified to:

```python
for param in target.parameters():
    if param is not None:
        return param.dtype
raise ValueError("All parameters in the model are None")
```

Also, when can `param` actually be `None`?

BenjaminBossan commented on src/peft/tuners/adamix.py:
```diff
@@ -0,0 +1,416 @@
+                hidden_states, other_args = hidden_states[0], hidden_states[1:]
+            hidden_states = f_out(hidden_states)
+            if tuple_input:
+                return (hidden_states,) + other_args
```
This could break downstream code if the type is a subtype of tuple and not exactly tuple. E.g. if the type is a `namedtuple`, `isinstance(hidden_states, tuple` would be `True` but the output would be a different type.

rishabbala replied:
That is one of the problems with the AdaMix model. Unlike Lora, where you can replace any layer with new low rank layers, adamix can only add the adapter after the selected layer. As a result, without knowing the exact class/function definition of the model, AdaMix is restricted. But I think this can be addressed in another PR

BenjaminBossan commented on src/peft/tuners/adamix.py:
```diff
@@ -0,0 +1,416 @@
+)
+
+
+if is_bnb_available():
+    pass
```
Remove?

BenjaminBossan commented on docs/source/conceptual_guides/adamix.mdx:
```diff
@@ -0,0 +1,53 @@
+
+
+<!--
+To make fine-tuning more efficient, IA3 (Infused Adapter by Inhibiting and Amplifying Inner Activations) 
```
I guess this should be replaced by a text about AdaMix ;)

BenjaminBossan commented on src/peft/tuners/adamix.py:
```diff
@@ -0,0 +1,416 @@
+        num_expert (`int`): The number of exprts per adapter module
+        sharing_down (`bool`): If the weights of the downsampling adapters are shared in each layer
+        sharing_up (`bool`): If the weights of the upsampling adapters are shared in each layer
+        return_two_views (`bool`): If two stochastic forward passes have to be made, and both outputs returned
```
Maybe note that this is only necessary for consistency training.

BenjaminBossan commented on src/peft/tuners/adamix.py:
```diff
@@ -0,0 +1,416 @@
+
+    Args:
+        target_modules (`Union[List[str],str]`): The names of the modules to apply AdaMix to.
+        adapter_dim (`int`):
```
I wonder if it should be called `r` here as well, for consistency with LoRA?

BenjaminBossan commented on src/peft/tuners/adamix.py:
```diff
@@ -0,0 +1,416 @@
+        super().__init__()
+
+        self.disable_adapters = False
+        self.return_two_views = return_two_views
```
I wonder: Instead of having the user set `return_two_views` as a config parameter, could this decision be made based on whether the model is in training mode or eval mode? I think we only ever need it the two views during training, right?

rishabbala replied:
Correct. In eval mode, even if this flag is set to True, only one value is returned. The reason I had this flag is in case for some reason we only want one view during training

BenjaminBossan commented on src/peft/tuners/adamix.py:
```diff
@@ -0,0 +1,416 @@
+    This is the configuration class to store the configuration of a [`AdaMixModel`].
+
+    Args:
+        target_modules (`Union[List[str],str]`): The names of the modules to apply AdaMix to.
```
Please add a few words about what types of layers are amenable to this method. This makes it easier if users want to employ this method for a model that is not among the listed default models.

BenjaminBossan reviewed (commented):
Thanks a lot for this PR.

I have a couple of comments on the code. Before looking through those, however, I think it's worth discussing the general approach of implementing AdaMix. If we decide on some bigger changes there, the individual comments I made could be obsolete.

From reading the paper, I think that AdaMix actually encompasses two separate methods:

1. There is the mixture of experts approach. This means that we can have a few adaptations instead of a single one, and those are all trained together (choosing one at random for each forward pass IIUC). The training also adds a consistency loss, which means this requires the user to have a custom training loop.
2. There is the method of combining multiple adapters ("model soup") from https://arxiv.org/abs/2203.05482. There are 3 approaches described in the paper, uniform, greedy, and learned, of which only uniform is implemented here IIUC.

In essence, these two methods are orthogonal and AdaMix is basically just putting those two together. In this PR, the two methods are tightly coupled, which I think makes it less useful than it could be. Ideally, I would like to see the two methods implemented independently, so that they can be composed more freely.

Regarding 1., we already have support in peft for multiple adapters on a single model. Unfortunately, I don't think we can easily leverage that to implement MoE. Could it be possible to have a wrapper class that takes a `LoraModel` / `IA3Model` etc. and implements MoE on top of those? If we could achieve that, I think it would be a better approach because it means we can apply the MoE idea to different types of adapters, which is also mentioned by the paper authors. As is, MoE is tightly integrated with a single adapter type (which is basically LoRA IIUC, but doesn't re-use any of the LoRA code).

Regarding 2., I would love to see the "model soup" mixing implemented as an independent method. Right now, for LoRA, we already have some ways to combine different adapters (linear, SVD). Could we generalize this code so that users can easily choose between multiple ways of combining adapters? However, I believe that "uniform" model soup is just like linear mixing with equal weights (haven't read that paper in detail, so not sure)?

If we could manage to achieve all of this, I think this would end up being a much more powerful contribution, as it gives the users a lot more flexibility. If we think this cannot be achieved, I'm okay with proceeding with the approach proposed in this PR.

rishabbala commented:
@BenjaminBossan, I think you are correct:

1. If I understand correctly, we want totake as input a Lora/IA3 model and implement MoE on top of this. I think this is possible but would have to be implemented separately outside the AdaMix model files. But this might raise issues in part 2, how do we combine the weights of all these layers, with (most likely) different dimensions during inference? One way might be using a gating/learned function that combines the output hidden states instead of the weights. However, I think this might deviate too far from the original PR.

2. Although the  authors discuss learned function to combine weights, only a uniform/linear mixing is presented in thee results and in their code. Im open to adding a learnable combination of parameters here (or Any other method).

From my understanding the current PR is the very basic setting of AdaMix, and it might be better to add this features one by one instead of in a single PR, but Im open to other suggestions as well

BenjaminBossan commented:
Good points @rishabbala. Let's wait what the other maintainers' opinion is here.

rishabbala commented:
Hey guys, I just want to follow up to decide how we want to proceed with this PR

rishabbala commented on examples/conditional_generation/peft_ia3_seq2seq.ipynb:
```diff
@@ -2,7 +2,8 @@
  "cells": [
   {
    "cell_type": "code",
-   "execution_count": 12,
```
It was unintended. I reverted the code back to the original version in the repo, but github still shows it as changed files

rishabbala commented on src/peft/tuners/adamix.py:
```diff
@@ -0,0 +1,416 @@
+            ]
+        return peft_config
+
+    def zero_adapters(self):
```
Added tests, and changed the name

rishabbala commented on src/peft/tuners/adamix.py:
```diff
@@ -0,0 +1,416 @@
+                    p_name = "weight"
+                else:
+                    p_name = "bias"
+                self.parameter_dict[p_name] = self.parameter_dict[p_name] + s_param / self.num_local_experts
```
Thats correct. Although the paper mentions other methods, their results and code contains only the uniform combination

rishabbala commented on tests/test_decoder_models.py:
```diff
@@ -172,10 +172,15 @@ def test_training_prompt_learning_tasks(self, test_name, model_id, config_cls, c
         self._test_disable_adapter(model_id, config_cls, config_kwargs)
+
+    @parameterized.expand(PeftTestConfigManager.get_grid_parameters(FULL_GRID))
+    def test_output_pair(self, test_name, model_id, config_cls, config_kwargs):
```
Added for all models

rishabbala commented:
@BenjaminBossan @younesbelkada I've addressed most of the issues raised so far. The stable diffusion test for LORA seems to be failing locally, but I haven't made changes in those files, so I think the error might be elsewhere. Let me know if there are any changes required

pacman100 reviewed (commented):
Hello, Thank you @rishabbala for the incredible work wrt adding AdaMix model which is a super cool method. Can you clarify one important thing:  When is the forward of `ExpertSoup` layer called? I see it being added via `target.add_module` but don't find it being called post the target's forward call. 

Left few comments. Post that good to merge. Thank you!

rishabbala commented:
Hey @pacman100 , thanks for the review. The forward of `ExpertSoup` is called when `forward` of the base model is called. Unlike LORA where you replace a selected layer with two low rank matrices, in this case, we add two linear projections after the selected layer. As a result the selected layer's weights do not change, and the weights of the adapter modules are called when `model.forward()` is called. Hope that makes sense

abhinav-kashyap-asus commented:
Hi guys, is there any update on this?? Seems to be stale from a couple of months
It could be great if AdaMix is added to PEFT. 
This can also serve as a basis for Mixture of Experts (MoE)
More and more papers are coming out for MoE and there is no unified codebase to do this properly 
peft can serve this gap! Happy to help. Let me know
